README
Perl5 extension module for authentication using tacacs+ server.

Copyright (C) 1998-99 Mike Shoyher <msh@corbina.net>, <msh@apache.lexa.ru>

tacacs+ is Copyright (c) 1995-1996 by Cisco systems, Inc. 

Thanks to:

	Mike McCauley <mikem@open.com.au>

This library is free software; you can redistribute it and/or
modify it under the same terms as Perl itself.








Changes
Revision history for Perl extension Authen::TacacsPlus.

0.01  Mon Dec  8 16:05:51 1997
	- original version; created by h2xs 1.18

0.16  Mon Apr 26 10:22:22 1999
	- Now does PAP and CHAP authentication, as well
	as the original ASCII. PAP and CHAP use TACACS+ version 1
	- Added flag to make_auth to specify the
	authentication type, plus string lengths (because
	in some auth types, the password string may contain NULs.
	- Added optional flag to ->authen() to specify the
	authentication type. Defaults to ASCII
	- Some reformatting to make it easier to read.
	- Cleaned up Makefiles so it builds cleaner
	- Mike McCauley (mikem@open.com.au)

0.17 Sat Sep 22 2007, Mike McCauley
     - Mike McCauley (mikem@open.com.au) is now co-maintainer.
     - DISTNAME changed to Authen-TacacsPlus to be consistent with file
     naming of other Authen modules.
     - readme changed to README.
     - Added licensing statement to README, with the approval and consent of
     Mike Shoyher.
     - Fixed a typedef in md5.h that breaks md5 on 64 bit machines. Patch
     provided by Ernst Oudhof.
     - Removed bogus broken constant TACACS_CLIENT.
     - Fixed a number of compiler warnings, but still does not compile on
     Windows.

0.18 Sat Mar 15 2008, Mike McCauley
     - Disabled default setting of -symbolic in Makefile.PL
     on Solaris since it seems to be
     no longer required on most platforms.
     - Improved errmsg() so it works as advertised
     - test suite now uses Test.
     - test suite now skips if there is no Tacacs server to test against

0.19 Sun Mar 16 2008, Mike McCauley
     - Fixed 'make test' failure in taclibplus due to missing pure_all target

0.20 Fri July 10, 2009 Mike McCauley
     - Fixed incorect spelling of 'Authentication' in tac_client.c
     - implements a 
     timeout for the connect() operation. It uses the timeout value which is
     passed to the init_tac_session() function. Contributed by Robert Leibl

0.21 Sat Aug 13, 2011 Mike McCauley
     - Changes to TacacsPlus.pm to permit multiple servers to be specified in
     new(). Patches provided by Paulo A Ferreira.
     
0.22 Wed Jan 18, 2012 Mike McCauley
     - Fixed warning under perl 5.14g

0.23 Wed Aug 23, 2012 Mike McCauley
     - Fixed problems in low level read_data() function triggered when an
     incorrect key is used with some Tacacs+ servers, resulting in a 0-length
     read(), causing a seg
     fault on some platforms, and a very slow exit on others. This problem
     appears to have been in tac_client ever since I inherited this library.

0.24 Fri Mar 22, 2013 Mike McCauley
     - Updated author and distribution location details to airspayce.com


test.pl
# Before `make install' is performed this script should be runnable with
# `make test'. After `make install' it should work as `perl test.pl'

######################### We start with some black magic to print on failure.

# Change 1..1 below to 1..last_test_to_print .
# (It may become useful if the test is moved to ./t subdirectory.)

BEGIN { $| = 1; print "1..7\n"; }
END {print "not ok 1\n" unless $loaded;}
use Authen::TacacsPlus;
$loaded = 1;
print "ok 1\n";

######################### End of black magic.

# Insert your test code below (better if it prints "ok 13"
# (correspondingly "not ok 13") depending on the success of chunk 13
# of the test code):

# You will have to change these to suit yourself:
my $host = 'zulu.open.com.au';
my $key = 'mysecret';
my $timeout = 15;
my $port = 49;
my $username = 'mikem';
my $password = 'fred';
# This is the CHAP encrypted password, including the challenge
# and identifier
my $chap_password = 'djfhafghlkdlkfjasgljksgljkdsjsdfshdfgsdfkjglgh';

my $tac = new Authen::TacacsPlus(Host=>$host,
				 Key=>$key,
				 Timeout=>$timeout,
				 Port=>$port);
if ($tac)
{
    print "ok 2\n";
}
else
{
    print "Could not connect to TACACSPLUS Host $self->{Host}: " . Authen::TacacsPlus::errmsg() . "\n";
    print "not ok 2\n" ;
}


# test default type (ASCII), backwards compatible
if ($tac->authen($username, $password))
{
    print "ok 3\n";
}
else
{
    my $reason = Authen::TacacsPlus::errmsg();
    print "authen failed: $reason\n";
    print "not ok 3\n" 
}
$tac->close();

my $tac = new Authen::TacacsPlus(Host=>$host,
				 Key=>$key,
				 Timeout=>$timeout,
				 Port=>$port);
if ($tac)
{
    print "ok 4\n";
}
else
{
    print "Could not connect to TACACSPLUS Host $self->{Host}: " . Authen::TacacsPlus::errmsg() . "\n";
    print "not ok 4\n" ;
}


# test default PAP type
if ($tac->authen($username, $password, &Authen::TacacsPlus::TAC_PLUS_AUTHEN_TYPE_PAP))
{
    print "ok 5\n";
}
else
{
    my $reason = Authen::TacacsPlus::errmsg();
    print "authen failed: $reason\n";
    print "not ok 5\n" 
}
$tac->close();

$tac = new Authen::TacacsPlus(Host=>$host,
				 Key=>$key,
				 Timeout=>$timeout,
				 Port=>$port);
if ($tac)
{
    print "ok 6\n";
}
else
{
    print "Could not connect to TACACSPLUS Host $self->{Host}: " . Authen::TacacsPlus::errmsg() . "\n";
    print "not ok 6\n" ;
}

# test CHAP auth type
require Digest::MD5;
$chap_id = '5';
$chap_challenge = '1234567890123456';
# This is the CHAP response from the NAS. We will fake it here
# by calculating it in the same way th eNAS does:
$chap_response = Digest::MD5::md5($chap_id . $password . $chap_challenge);
$chap_string = $chap_id . $chap_challenge . $chap_response;

if ($tac->authen($username, $chap_string, &Authen::TacacsPlus::TAC_PLUS_AUTHEN_TYPE_CHAP))
{
    print "ok 7\n";
}
else
{
    my $reason = Authen::TacacsPlus::errmsg();
    print "authen failed: $reason\n";
    print "not ok 7\n" 
}
$tac->close();



tacpluslib/Makefile.PL
use ExtUtils::MakeMaker;                                         
$Verbose = 1;                                                    
use Config;

$osname=$Config{osname};
%defines=(
        'linux'=>'-DLINUX',
        'freebsd'=>'-DFREEBSD',
        'netbsd'=>'-DNETBSD',
        'solaris'=>'-DSOLARIS',
        'sunos'=>'-DSOLARIS',
        'bsdi'=>'-DBSDI',
        'aix'=>'-DAIX',
        'hpux'=>'-DHPUX',
);


WriteMakefile(                                                   
    NAME      => 'Authen::TacacsPlus::tacplus',                               
    DEFINE    => $defines{$osname},
    SKIP      => [qw(static dynamic)],
    clean     => {'FILES' => 'libtacplus$(LIB_EXT)'},              
);                                                               
                                                                 
sub MY::top_targets {     

'
all:	static                                   
                                                 
static:	libtacplus$(LIB_EXT)              
                                                 
libtacplus$(LIB_EXT):	$(O_FILES)                  
	$(AR) cr libtacplus$(LIB_EXT) $(O_FILES)  
	$(RANLIB) libtacplus$(LIB_EXT)  

pure_all : 
	$(NOECHO) $(NOOP)

                                                 
 ';                                              
 }                                               



tacpluslib/tac_client.c

/* (C) 1997 Mike Shoyher msh@corbina.net, msh@apache.lexa.ru */

#include<netdb.h>
#include<stdio.h>
#include<netinet/in.h>
#include<sys/socket.h>
#include<arpa/inet.h>
#include<string.h>
#include<stdlib.h>
#include <sys/types.h> 
#include <sys/time.h>  
#include <unistd.h>  
#include <time.h>  
#include"tac_plus.h"
#include "tacplus_client.h"

int tac_sequence;
int tac_session_id;
int tac_maxtry=3;
int tac_timeout=15;
int tac_fd;                  
char tac_key[128];
struct sockaddr_in tac_port; 
struct hostent *tac_h;       
struct servent *tac_serv;    
char ourhost[128];
int ourhost_len;
char* ourtty="Virtual00";
int ourtty_len;
char* tac_err="NONE";

void fill_tac_hdr(struct tac_plus_pak_hdr* hdr)
{
    hdr->version=TAC_PLUS_VER_0;   
    hdr->type=TAC_PLUS_AUTHEN;     
    hdr->seq_no=tac_sequence;      
    hdr->encryption=TAC_PLUS_CLEAR;
    hdr->session_id=tac_session_id;
}

int send_data(void* buf, int buf_len, int fd);
int read_reply(void** data);

/* Need to specify the lengths, becuase CHAP passwords etc 
   may have NULs in them */
void send_auth_cont(char* msg, int msg_len);
int make_auth(char* username, int user_len,
	      char* password, int password_len,
	      int authen_type)
{
    struct tac_plus_pak_hdr hdr;
    struct authen_start as;
    int datalength;
    void* buf;
    void* data;
    int data_len;
    int buf_len;
    struct authen_reply* ar;

    fill_tac_hdr(&hdr);

    datalength=TAC_AUTHEN_START_FIXED_FIELDS_SIZE;


    as.action= TAC_PLUS_AUTHEN_LOGIN ;
    as.priv_lvl= TAC_PLUS_PRIV_LVL_MIN ;
    as.authen_type = authen_type ;
    as.service= TAC_PLUS_AUTHEN_SVC_LOGIN ;
    as.user_len=as.port_len=as.rem_addr_len=as.data_len=0;

    if (authen_type != TAC_PLUS_AUTHEN_TYPE_ASCII)
    {
	/* This will be a version 1 request with the username
	   and password in the start */
	hdr.version=TAC_PLUS_VER_1;
	as.user_len = user_len;
	as.data_len = password_len;
	
    }
    buf=malloc(buf_len=TAC_PLUS_HDR_SIZE+TAC_AUTHEN_START_FIXED_FIELDS_SIZE+ourtty_len+ourhost_len+as.user_len+as.data_len);

    /* Append user name if required */
    bcopy(username,buf+datalength+TAC_PLUS_HDR_SIZE,as.user_len);
    datalength+=(as.user_len);

    bcopy(ourtty,buf+datalength+TAC_PLUS_HDR_SIZE,ourtty_len);
    datalength+=(ourtty_len);
    as.port_len=ourtty_len;

    bcopy(ourhost,buf+datalength+TAC_PLUS_HDR_SIZE,ourhost_len);      
    datalength+=(ourhost_len); 
    as.rem_addr_len=ourhost_len;

    /* Append password if required */
    bcopy(password,buf+datalength+TAC_PLUS_HDR_SIZE,as.data_len);
    datalength+=(as.data_len);

    hdr.datalength= htonl(datalength) ;
    bcopy(&hdr,buf,TAC_PLUS_HDR_SIZE);
    bcopy(&as,buf+TAC_PLUS_HDR_SIZE,TAC_AUTHEN_START_FIXED_FIELDS_SIZE);
    md5_xor(buf, buf+TAC_PLUS_HDR_SIZE, tac_key);
    send_data(buf,buf_len,tac_fd);
    free(buf);
    while ((data_len=read_reply(&data))!=-1){

	ar=data;
	switch (ar->status) {
	case TAC_PLUS_AUTHEN_STATUS_GETUSER:
	    free(data);
	    send_auth_cont(username, user_len);
	    break;
	case TAC_PLUS_AUTHEN_STATUS_GETPASS:
	    free(data);
	    send_auth_cont(password, password_len);   
	    break;                      
	case TAC_PLUS_AUTHEN_STATUS_PASS:
	    return 1;
	case TAC_PLUS_AUTHEN_STATUS_FAIL: 
	    tac_err="Authentication failed";
	    return 0;                 
	default :
	    tac_err="Protocol error";
	    return 0;
	}

    }
    tac_err="Unknown error";
    return 0;
}

void send_auth_cont(char* msg, int msg_len)
{
struct tac_plus_pak_hdr hdr;
struct authen_cont ac;     
int datalength;             
void* buf;                  
int buf_len;

buf_len=TAC_PLUS_HDR_SIZE+TAC_AUTHEN_CONT_FIXED_FIELDS_SIZE+msg_len;
buf=malloc(buf_len);
fill_tac_hdr(&hdr);
ac.user_data_len=ac.flags=0;
ac.user_msg_len=htons(msg_len);
bcopy(msg,buf+TAC_PLUS_HDR_SIZE+TAC_AUTHEN_CONT_FIXED_FIELDS_SIZE,msg_len); 
datalength=msg_len+TAC_AUTHEN_CONT_FIXED_FIELDS_SIZE;                                  
hdr.datalength=htonl(datalength);
bcopy(&hdr,buf,TAC_PLUS_HDR_SIZE);                                    
bcopy(&ac,buf+TAC_PLUS_HDR_SIZE,TAC_AUTHEN_CONT_FIXED_FIELDS_SIZE);  
md5_xor(buf, buf+TAC_PLUS_HDR_SIZE, tac_key);                         
send_data(buf,buf_len,tac_fd);                                        
free(buf);
}

void myerror (char* s)
{
printf("%s\n",s);
exit(1);
}

int init_tac_session(char* host_name, char* port_name, char* key, int timeout)
{

int flags;
int res;
int optval;
socklen_t len;
fd_set wset;
struct timeval tv;

gethostname(ourhost,127);
ourhost_len=strlen(ourhost);
ourtty_len=strlen(ourtty);
srand(time(NULL));
if (timeout>0) tac_timeout=timeout;
strcpy(tac_key,key);
tac_session_id=rand();
tac_sequence=1;
tac_port.sin_family = AF_INET; 
if (*host_name>='0' && *host_name<='9') {                              
     tac_port.sin_addr.s_addr = inet_addr(host_name);                     
} else {                                                                
     tac_h = gethostbyname(host_name);                                    
     if (!tac_h) {                                                         
	 /*myerror("Cannot resolve host name");*/
	 tac_err="Cannot resolve host name";
         return -1;                                                   
     }                                                              
     tac_port.sin_addr = *((struct in_addr *) tac_h->h_addr);                 
}                                                                  
if (port_name==NULL) port_name="tacacs";
if (*port_name>='0' && *port_name<='9') {
  tac_port.sin_port=htons (atoi(port_name));
} else {
    tac_serv=getservbyname(port_name,"tcp");
    if (tac_serv)
      tac_port.sin_port=tac_serv->s_port;
    else {
        /*myerror("Unknown port");*/
	tac_err="Unknown port";
        return -1;
    }
}
if((tac_fd = socket (AF_INET, SOCK_STREAM, 0))<0) return -1;

// get flags
flags = fcntl(tac_fd, F_GETFL, 0);
if( flags < 0 ) {
	//fprintf( stderr, "fcntl: %s\n", strerror(errno) );
	tac_err = "socket error";
	return -1;
}

// set socket to nonblock
res = fcntl( tac_fd, F_SETFL, flags | O_NONBLOCK );
if( res < 0 ) {
	//fprintf( stderr, "fcntl: %s\n", strerror(errno) );
	tac_err = "socket error";
	return -1;
}

// connect
res = connect (tac_fd, (struct sockaddr *) &tac_port, sizeof tac_port);

// connection not established, but in progress
if( res < 0 && (errno != EINPROGRESS) ) {
	tac_err = "connection failed";
	return -1;
}

// wait for connection or timeout
if( res != 0 ) {
	FD_ZERO(&wset);
	FD_SET(tac_fd, &wset);
	tv.tv_sec  = timeout;
	tv.tv_usec = 0;

	res = select( tac_fd+1, NULL, &wset, NULL, &tv );
	if( res < 0 ) {
		tac_err = "select failed";
		return -1;
	}
	else if( res == 0 ) {
		tac_err = "timeout";
		return -1;
	}
	if( res > 0 ) {
		// socket is ready for writing. Check if connection was
		// established successfully
		len = sizeof(optval);
		if( getsockopt( tac_fd, SOL_SOCKET, SO_ERROR, (void *)&optval, &len ) > 0 ) {
			tac_err = "getsockopt failed";
			return -1;
		}
		if( optval != 0 ) {
			tac_err = "connection failed";
			return -1;
		}
		// optval == 0 --> no error, connection established
	}
}
	
return tac_fd;
}

int send_data(void* buf, int buf_len, int fd)
{
fd_set fds;
int rr;
int try;
struct timeval tv;

FD_ZERO(&fds);
FD_SET(fd,&fds);
tv.tv_sec = tac_timeout;
tv.tv_usec = 0;     
for(try=0;try<tac_maxtry;try++){
	rr=select(fd+1,NULL,&fds,NULL,&tv);
	if (FD_ISSET(fd,&fds)) {
        	if (buf_len!=write(fd,buf,buf_len)) continue;
		return 0;
	}
	myerror("Write error");
	return -1;
}
 return 1;
}

int read_data(void* buf, int buf_len, int fd)                 
{                                                             
fd_set fds;                                                   
int rr;                                                       
int try;                                                      
struct timeval tv;                                            
                                                              
FD_ZERO(&fds);                                                
FD_SET(fd,&fds);                                              
tv.tv_sec = tac_timeout;                                      
tv.tv_usec = 0;                                               
for(try=0;try<tac_maxtry;try++){                              
        rr=select(fd+1,&fds,NULL,NULL,&tv);                   
        if (FD_ISSET(fd,&fds)) {                              
                if (buf_len!=read(fd,buf,buf_len)) continue; 
                return 0;                                     
        }                                                     
        myerror("read error");                               
        return -1;                                            
}                                                             
 return 1;                                                              
}                                                             

int read_reply(void** data)
{
int data_len;
struct tac_plus_pak_hdr hdr;
if (read_data(&hdr,TAC_PLUS_HDR_SIZE,tac_fd)==-1) return -1;
data_len=ntohl(hdr.datalength);
tac_sequence=hdr.seq_no+1;
*data=malloc(data_len);
if (read_data(*data,data_len,tac_fd)==-1) return -1;
md5_xor(&hdr, *data, tac_key); 
return data_len;
}

void deinit_tac_session()
{
	shutdown(tac_fd,2);
	close(tac_fd);
}


void report()
{
}

int debug;


TacacsPlus.pm
#
# (C) 1998 Mike Shoyher <msh@corbina.net>, <msh@apache.lexa.ru>

package Authen::TacacsPlus;

use strict;
use Carp;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK $AUTOLOAD);

require Exporter;
require DynaLoader;

@ISA = qw(Exporter DynaLoader);
# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.
@EXPORT_OK = qw(
	TACPLUS_CLIENT
);
$VERSION = '0.24';

sub new
{
my $class = shift;
my %h;
my $self = {};
bless $self, $class;
$self->{'servers'} = [];
if (ref $_[0] eq 'ARRAY') {
    %h = @{ $_[0] };
    shift @_;
    push @{ $self->{'servers'} }, @_;
} else {
    %h = @_;
}
my $res=-1;
$self->{'timeout'} = $h{'Timeout'} ? $h{'Timeout'} : 15;
$self->{'port'} = $h{'Port'} ? $h{'Port'} : 'tacacs';
$self->{'host'} = $h{'Host'};
$self->{'key'} = $h{'Key'};
$res=init_tac_session($self->{'host'},$self->{'port'},
	$self->{'key'},$self->{'timeout'});
if ($res<0) {
    my $s = $self->{'servers'};
    while ($s->[0]) {
        my %h = @{ $s->[0] };
        shift @{ $s };
        $res=init_tac_session( $h{'Host'},
                               $h{'Port'} ? $h{'Port'} : 'tacacs',
                               $h{'Key'},
                               $h{'Timeout'} ? $h{'Timeout'} : 15
                              );
        last if ($res >= 0);
    }
}
undef $self if ($res < 0);
$self;
}

# Third arg authen_type is optional, defaults to 
# TAC_PLUS_AUTHEN_TYPE_ASCII
sub authen
{
    my $self = shift;
    my $username = shift;
    my $password = shift;
    my $authen_type = shift || &Authen::TacacsPlus::TAC_PLUS_AUTHEN_TYPE_ASCII;
    my $res=make_auth($username,$password,$authen_type);
    unless ($res || errmsg() =~ /Authentication failed/) {
        my $s = $self->{'servers'};
        while ($s->[0]) {
            my %h = @{ $s->[0] };
            shift @{ $s };
            my $ret=init_tac_session( $h{'Host'},
                                      $h{'Port'} ? $h{'Port'} : 'tacacs',
                                      $h{'Key'},
                                      $h{'Timeout'} ? $h{'Timeout'} : 15
                                    );
            next if ($ret < 0);
            $res=make_auth($username,$password,$authen_type);
            last if $res;
        }

    }
    $res;
}

sub close
{
    deinit_tac_session();
}



sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.  If a constant is not found then control is passed
    # to the AUTOLOAD in AutoLoader.

    my $constname;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    my $val = constant($constname, @_ ? $_[0] : 0);
    if ($! != 0) {
	if ($! =~ /Invalid/) {
	    $AutoLoader::AUTOLOAD = $AUTOLOAD;
	    goto &AutoLoader::AUTOLOAD;
	}
	else {
		croak "Your vendor has not defined Authen::TacacsPlus macro $constname";
	}
    }
    eval "sub $AUTOLOAD { $val }";
    goto &$AUTOLOAD;
}

bootstrap Authen::TacacsPlus $VERSION;

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__
# Below is the stub of documentation for your module. You better edit it!

=head1 NAME

Authen::TacacsPlus - Perl extension for authentication using tacacs+ server

=head1 SYNOPSIS

  use Authen::TacacsPlus;

  $tac = new Authen::TacacsPlus(Host=>$server,
			Key=>$key,
			[Port=>'tacacs'],
			[Timeout=>15]);

  or

  $tac = new Authen::TacacsPlus(
     [ Host=>$server1, Key=>$key1, [Port=>'tacacs'], [Timeout=>15] ],
     [ Host=>$server2, Key=>$key2, [Port=>'tacacs'], [Timeout=>15] ],
     [ Host=>$server3, Key=>$key3, [Port=>'tacacs'], [Timeout=>15] ],
     ...  );

  $tac->authen($username,$passwords);

  Authen::TacacsPlus::errmsg(); 

  $tac->close();


=head1 DESCRIPTION

Authen::TacacsPlus allows you to authenticate using tacacs+ server.

  $tac = new Authen::TacacsPlus(Host=>$server,      
 	                Key=>$key,          
                        [Port=>'tacacs'],   
                        [Timeout=>15]);     

Opens new session with tacacs+ server on host $server, encrypted
with key $key. Undefined object is returned if something wrong
(check errmsg()).

With a list of servers the order is relevant. It checks the availability
of the Tacacs+ service using the order you defined.


  Authen::TacacsPlus::errmsg();

Returns last error message.  

  $tac->authen($username,$password,$authen_type);

Tries an authentication with $username and $password. 1 is returned if
authenticaton succeded and 0 if failed (check errmsg() for reason).

$authen_type is an optional argument that specifies what type
of authentication to perform. Allowable options are:
Authen::TacacsPlus::TAC_PLUS_AUTHEN_TYPE_ASCII (default)
Authen::TacacsPlus::TAC_PLUS_AUTHEN_TYPE_PAP
Authen::TacacsPlus::TAC_PLUS_AUTHEN_TYPE_CHAP

ASCII uses Tacacs+ version 0, and will authenticate against 
the "login" or "global" password on the Tacacs+ server. If no
authen_type is specified, it defaults to this type of authentication.

PAP uses Tacacs+ version 1, and will authenticate against 
the "pap" or "global" password on the Tacacs+ server.

CHAP uses Tacacs+ version 1, and will authenticate against 
the "chap" or "global" password on the Tacacs+ server. With CHAP,
the password if formed by the concatenation of
  chap id + chap challenge + chap response

There is example code in test.pl

If you use a list of servers you can continue using $tac->authen if
one of them goes down or become unreachable.


  $tac->close();

Closes session with tacacs+ server.

=head1 EXAMPLE

  use Authen::TacacsPlus;                                             
                                                              
                                                              
  $tac = new Authen::TacacsPlus(Host=>'foo.bar.ru',Key=>'9999');  
  unless ($tac){                                              
          print "Error: ",Authen::TacacsPlus::errmsg(),"\n";          
          exit(1);                                            
  }                                                           
  if ($tac->authen('john','johnpass')){                   
          print "Granted\n";                                  
  } else {                                                    
          print "Denied: ",Authen::TacacsPlus::errmsg(),"\n";         
  }                                                           
  $tac->close();                                              
  


=head1 AUTHOR

Mike Shoyher, msh@corbina.net, msh@apache.lexa.ru

Mike McCauley, mikem@airspayce.com

=head1 BUGS

only authentication is supported

only one session may be active (you have to close one session before
opening another one)

=head1 SEE ALSO

perl(1).

=cut
